### PR TITLE
Tennessee Mapping (tested)

### PR DESCRIPTION
**Pull request by guegueng**
first draft of Tennessee mapping. several questions

### CODE REVIEW

File: heidrun/TN_mods.rb [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+  preview :class => DPLA::MAP::WebResource do
+    uri record.field('mods:location', 'mods:url')
+              .match_attribute(:access, 'preview')
+  end
+
+  originalRecord :class => DPLA::MAP::WebResource do
+    uri record_uri
+  end
+#they have a mapping for format and rights in the Web Resource as well...not sure how to implement
```
Comment (anarchivist) [inline heidrun/TN_mods.rb]: Can you share an example? I'm guessing this would be for `preview`, or maybe `object`.

File: heidrun/TN_mods.rb [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    genre record.field('mods:genre').match_attribute(:authority 'aat')
+
+    identifier record.field('mods:identifier')
+
+    language :class => DPLA::MAP::Controlled::Language,
+             :each => record.field('mods:language', 'mods:languageTerm')
+             				.match_attribute(:type 'code')
+             				.match_attribute(:authority 'iso639-2b'),
+             :as => :lang do
+      prefLabel lang
```
Comment (anarchivist) [inline heidrun/TN_mods.rb]: Should be `providedLabel`

File: heidrun/TN_mods.rb [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    end
+
+    spatial :class => DPLA::MAP::Place,
+            :each => record.field('mods:subject', 'mods:geographic'),
+            :as => :place do
+      providedLabel place
+    end
+    
+    #They have coordinates as well. I wasn't sure if I should add a second spatial mapping or include this in the previous. I decided to add a second. Is that correct?
+    spatial :clas => DPLA::MAP::Place,
```
Comment (anarchivist) [inline heidrun/TN_mods.rb]: typo - should be `:class`

File: heidrun/TN_mods.rb [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+      prefLabel lang
+    end
+
+    spatial :class => DPLA::MAP::Place,
+            :each => record.field('mods:subject', 'mods:geographic'),
+            :as => :place do
+      providedLabel place
+    end
+    
+    #They have coordinates as well. I wasn't sure if I should add a second spatial mapping or include this in the previous. I decided to add a second. Is that correct?
```
Comment (anarchivist) [inline heidrun/TN_mods.rb]: I think you'd probably want to use `record.fields(...)`, such as in [here](https://github.com/dpla/heidrun-mappings/blob/4fd19163f15b91167f16abb77390c67f5a7fdabb/heidrun/uw_qdc.rb#L34)

File: heidrun/TN_mods.rb [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      wgs84_pos:lat lat
+    end
+
+    publisher :class => DPLA::MAP::Agent,
+              :each => record.field('mods:originInfo'),
+              :as => :publisher do
+      providedLabel publisher.field('mods:publisher')
+    end
+    
+    #totally a guess as to how to not map properties with certain attributes. for relation
```
Comment (anarchivist) [inline heidrun/TN_mods.rb]: See here: https://github.com/dpla/heidrun-mappings/blob/master/heidrun/esdn_mods.rb#L58-L59

File: heidrun/TN_mods.rb [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    relation record.field('mods:relatedItem')
+    				.reject_attribute(:type 'isReferencedBy')
+    				.reject_attribute(:type 'references')
+    				.field('mods:titleInfo', 'mods:title')
```
Comment (anarchivist) [inline heidrun/TN_mods.rb]: Do `relation`, `isReplacedby`, and `replaces` need to be _either_ the `mods:title` or `mods:url`, or _both_?

File: heidrun/TN_mods.rb (hunk 1 above)
Comment (guegueng) [inline heidrun/TN_mods.rb]: I'm thinking its for object (which they also have a mapping for...I did not include that here either). 
Object = location/url access="raw object"
format = physicalDescription/internetMediaType
rights = accessCondition displayLabel="Digital Object Rights"

File: heidrun/TN_mods.rb (hunk 2 above)
Comment (guegueng) [inline heidrun/TN_mods.rb]: right. will fix.

File: heidrun/TN_mods.rb (hunk 4 above)
Comment (guegueng) [inline heidrun/TN_mods.rb]: ok.

File: heidrun/TN_mods.rb (hunk 5 above)
Comment (guegueng) [inline heidrun/TN_mods.rb]: I saw that, but I wasn't sure if it was the same scenario. In the example, you are disregarding all "point" attributes. In this case I want to only reject those with the value "isReferencedBy" or "references"

File: heidrun/TN_mods.rb (hunk 6 above)
Comment (guegueng) [inline heidrun/TN_mods.rb]: both

**Comment by no-reply** [comment]
An example of this mapping run over a record is at https://gist.github.com/no-reply/458a8d09e8d622c831a7

**Comment by no-reply** [comment]
As noted in inline comments, this can be simplified somewhat with merge and deployment of https://github.com/dpla/KriKri/pull/175

File: heidrun/TN_mods.rb [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    label 'Tennessee Digital Library'
+  end
+
+  dataProvider :class => DPLA::MAP::Agent do
+    providedLabel record.field('mods:recordInfo', 'mods:recordContentSource')
+  end
+
+  isShownAt :class => DPLA::MAP::WebResource do
+    uri record.field('mods:location', 'mods:url')
+         .match_attribute(:usage, 'primary display')
```
Comment (no-reply) [inline heidrun/TN_mods.rb]: Should this be `:usage, 'primary'`, as seen at: https://gist.github.com/no-reply/458a8d09e8d622c831a7#file-record-xml-L44 ?

Comment (anarchivist) [inline heidrun/TN_mods.rb]: Yes, this appears the case, per the [crosswalk document](https://docs.google.com/spreadsheets/d/1BzZvDOf4fgas3TD21xF40lu2pk2XW0k2pTGJKIt6438/edit#gid=1453046017).

Comment (no-reply) [inline heidrun/TN_mods.rb]: Fixed in 1b53fd6.

**Comment by no-reply** [comment]
This is squashed.

**Pull request merged**